File: frontend/utilities/scripts_encoding.ts
// Header name for signaling base64-encoded scripts to bypass WAF rules
// that may block requests containing shell/PowerShell script patterns.
export const SCRIPTS_ENCODED_HEADER = "X-Fleet-Scripts-Encoded";

/**
 * Base64 encode a script string with proper UTF-8 handling.
 * Returns undefined for undefined input and empty string for empty input,
 * which allows callers to pass through empty/unset script fields without modification.
 */
export const encodeScriptBase64 = (
  script: string | undefined
): string | undefined => {
  if (script === undefined) {
    return undefined;
  }
  if (script === "") {
    return "";
  }
  // Use TextEncoder for proper UTF-8 handling of unicode characters
  const encoder = new TextEncoder();
  const data = encoder.encode(script);
  // Convert Uint8Array to binary string, then base64
  let binary = "";
  data.forEach((byte) => {
    binary += String.fromCharCode(byte);
  });
  return btoa(binary);
};

File: server/service/scripts_encoding.go
package service

import (
	"encoding/base64"
	"net/http"
)

// ScriptsEncodedHeader is the HTTP header used to signal that script fields
// in the request body are base64-encoded. This is used to bypass WAF rules
// that may block requests containing shell/PowerShell script patterns.
const ScriptsEncodedHeader = "X-Fleet-Scripts-Encoded"

// decodeBase64Script decodes a base64-encoded script string.
// Returns empty string for empty input, which allows callers to pass through
// empty/unset script fields without modification.
func decodeBase64Script(encoded string) (string, error) {
	if encoded == "" {
		return "", nil
	}
	decoded, err := base64.StdEncoding.DecodeString(encoded)
	if err != nil {
		return "", err
	}
	return string(decoded), nil
}

// isScriptsEncoded checks if the request has the scripts encoding header
// set to "base64", indicating that script fields should be decoded.
func isScriptsEncoded(r *http.Request) bool {
	return r.Header.Get(ScriptsEncodedHeader) == "base64"
}
